File: solutions/606.construct-string-from-binary-tree/construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* t) {
        if(!t) return "";
        string str(to_string(t->val));
        string lstr = tree2str(t->left);
        string rstr = tree2str(t->right);
        if (!lstr.empty() || !rstr.empty()) 
            str += "(" + lstr + ")";
        if (!rstr.empty()) 
            str += "(" + rstr + ")";
        return str;
    }
};

File: solutions/606.construct-string-from-binary-tree/construct-string-from-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public String tree2str(TreeNode t) {
        if(t == null) return "";
        StringBuilder res = new StringBuilder();
        res.append(t.val);
        if(t.left == null && t.right == null) return res.toString();
        res.append("(" + tree2str(t.left) + ")");
        if(t.right != null) 
            res.append("(" + tree2str(t.right) + ")");
        return res.toString();
    }
}